seccomp/umount.go

```diff
@@ -113,19 +120,237 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
 		return u.tracer.createSuccessResponse(u.reqId), nil
 	}
 
+	// Verify if the umount op is addressing an immutable resource and prevent
+	// it if that's the case.
+	if ok, resp := u.umountAllowed(mip); ok == false {
+		return resp, nil
+	}
+
 	// Not a mount we manage, have the kernel do the unmount.
 	return u.tracer.createContinueResponse(u.reqId), nil
 }
 
+// umountAllowed purpose is to prevent immutable resources from being
+// unmounted.
+//
+// Method will return 'true' when the unmount operation is deemed legit, and
+// will return 'false' otherwise.
+func (u *umountSyscallInfo) umountAllowed(
+	mip domain.MountInfoParserIface) (bool, *sysResponse) {
+
+	// Skip file-systems explicitly handled by sysbox-fs.
+	if u.FsType == "proc" || u.FsType == "sysfs" {
+		return true, nil
+	}
+
+	// There must be mountinfo state present for this target. Otherwise, let
+	// kernel handle the error.
+	info := mip.GetInfo(u.Target)
+	if info == nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	//
+	// The following scenarios are relevant within the context of this function
+	// and will be handled separately to ease the logic comprehension and its
+	// maintenability / debuggability.
+	//
+	// The different columns in this table denote the 'context' in which the
+	// unmount process is executing, and thereby, dictates the logic chosen
+	// to handle each unmount request.
+	//
+	//    +-----------+--------------+--------------+----------+
+	//    | Scenarios | Unshare(mnt) | Pivot-root() | Chroot() |
+	//    +-----------+--------------+--------------+----------+
+	//    | 1         | no           | no           | no       |
+	//    | 2         | no           | yes          | no       |
+	//    | 3         | no           | no           | yes      |
+	//    | 4         | no           | yes          | yes      |
+	//    | 5         | yes          | no           | no       |
+	//    | 6         | yes          | yes          | no       |
+	//    | 7         | yes          | no           | yes      |
+	//    | 8         | yes          | yes          | yes      |
+	//    +-----------+--------------+--------------+----------+
+	//
+
+	// Identify the mount-ns of the process launching the unmount to compare it
+	// with the one of the sys container's initpid. In the unlikely case of an
+	// error, let the kernel deal with it.
+	processMountNs, err := u.processInfo.MountNsInode()
+	if err != nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
```

Comment:
We should explicitly return an error here; we don't want to rely on the kernel doing this (what if it allows the unmount, then we would have a security hole).